torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -242,6 +251,135 @@ class BatchDataBuffer {
+/// during chunk loading.
+class ChunkSelector {
+ public:
+  virtual ~ChunkSelector() = default;
+  ChunkSelector(size_t chunk_count, size_t num_replicas = 1, size_t rank = 0)
+      : chunk_count_(chunk_count),
+        num_replicas_(num_replicas),
+        rank_(rank),
+        epoch_(0) {
+    local_chunk_count_ =
+        static_cast<size_t>(std::ceil(chunk_count_ * 1.0 / num_replicas_));
+  }
+
+  /// Get the next chunk index for loading.
+  /// Note: this method needs to be thread-safe.
+  virtual optional<size_t> next() = 0;
+
+  /// Reset the chunk selector for a new enumeration.
+  virtual void reset() = 0;
+
+  /// Set the epoch for the current enumeration. This can be used to alter the
+  /// chunk selection and shuffling behavior.
+  void set_epoch(size_t epoch) {
+    epoch_ = epoch;
+  }
+
+  /// Return the number of chunks to be loaded. In the case of distributed
+  /// training, this is different to chunk_count as each loader needs to load
+  /// only a subset of chunks.
+  size_t local_chunk_count() {
+    return local_chunk_count_;
+  }
+
+ protected:  
+  size_t chunk_count_;
+  size_t num_replicas_;
+  size_t rank_;
+  size_t epoch_;
+  size_t local_chunk_count_;
+};
+
+/// Select chunks randomly. The chunk order shuffled at each `reset()` call.
+class RandomChunkSelector : public ChunkSelector {
+ public:
+  RandomChunkSelector(
+      size_t chunk_count,
+      size_t num_replicas = 1,
+      size_t rank = 0)
+      : ChunkSelector(chunk_count, num_replicas, rank) {
+    size_t index_count =
+        num_replicas_ == 1 ? chunk_count_ : local_chunk_count_ * num_replicas_;
+    all_indices_.resize(index_count);
+    std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+    if (num_replicas_ > 1)
+    {
+      for (size_t i = chunk_count; i < index_count; ++i) {
+        all_indices_[i] =
+            i % chunk_count_; // we are adding some more chunks to make all
+                              // replicas to have the same number of chunks.
+      }    
+    }
+  }
+
+  optional<size_t> next() override {
+    AT_CHECK(
+        !chunk_indices_.empty(),
+        "reset() needs to be called before calling next().");
+    size_t idx = current_index_.fetch_add(1, std::memory_order_relaxed);
+    if (idx < chunk_indices_.size()) {
+      return chunk_indices_[idx];
+    } else {
+      return nullopt;
+    }
+  }
+
+  void reset() override {
+    std::minstd_rand rand(epoch_);
+    std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
+    chunk_indices_.clear();
+    chunk_indices_.reserve(local_chunk_count_);
+    auto begin_index_iter = all_indices_.begin() + rank_ * local_chunk_count_;
+    auto end_index_iter = begin_index_iter + local_chunk_count_;
+    std::copy(begin_index_iter, end_index_iter, back_inserter(chunk_indices_));
+    current_index_ = 0;
+  }
+
+ private:
+  std::vector<size_t> all_indices_;
+  std::vector<size_t> chunk_indices_;
```

Comment:
Depending on the size of `chunk_indices`,  we could discard it to save memory and time during vector copying from `all_indices_` to `chunk_indices_`

We can keep a single copy of indices at `all_indices_` and each replica would keep two iterators for the beginning (`rank_ * local_chunk_count_`) and ending (`rank_ * local_chunk_count_ + local_chunk_count_`) of its own chunks using. Indexing would be `rank_ * local_chunk_count_ + current_index_`

Reply:
Thank you for the suggestion. Fixed it.